### ISSUE TITLE
Selective Attention

### ISSUE THREAD

**Issue by kazunator** [lucidrains/x-transformers#278]
I see that the "Selective Attention Improves Transformer" paper has been cited, but I don't see any example or implementation. Did I miss it?

**Comment by lucidrains** [comment]
@kazunator it is available behind [`attn_selective = True`](https://github.com/lucidrains/x-transformers/blob/main/x_transformers/x_transformers.py#L931)

**Comment by lucidrains** [comment]
@kazunator logic is [here](https://github.com/lucidrains/x-transformers/blob/main/x_transformers/attend.py#L68) if you wish to review it

**Comment by kazunator** [comment]
Oh gotcha, thanks for the reply. Was the context pruning from this paper also implemented? They have an interesting way of doing it with the memory budget thing

**Comment by lucidrains** [comment]
@kazunator one other thing that should be explored is using the same scheme, but doing it depthwise (spelling it out, slowly focusing more attention on what matters as you traverse depth). just putting the idea out there

**Comment by kazunator** [comment]
@lucidrains Interesting, right now they just the first head for all layers and they've said that it worked best. I didn't completely understand what you meant by doing the same scheme depthwise, so I would love if you expanded more on it.
Right now I'm trying to reproduce the results of the paper (with some different settings, I used the norms and hyperparameters of nanogpt instead) and I'm trying to compare it with a normal transformer to see if I can get the same perplexity gains. Hopefully my implementation is correct, but if something goes wrong I'll just use yours.
After that, I'll work on context pruning and if that works, I'll see if I can fine-tune a llama 1b with selective attention. Hopefully, SFT is enough to see at least some gains in memory with context pruning

**Comment by kazunator** [comment]
This my implementation of selective attention for context:
```
    def forward(self, x):
        B, T, C = x.size() # batch size, sequence length, embedding dimensionality (n_embd)
        # calculate query, key, values for all heads in batch and move head forward to be the batch dim
        qkv = self.c_attn(x)
        q, k, v = qkv.split(self.n_embd, dim=2)
        k = k.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
        q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
        v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)

        # Standard attention computation
        att = (q @ k.transpose(-2, -1)) * (1.0 / math.sqrt(k.size(-1)))
        att = att.masked_fill(self.bias[:,:,:T,:T] == 0, float('-inf'))

        # Apply selective attention
        S = att[:, 0].clone()  # Select head 0 logits (clone to avoid in-place modification issues)
        S = F.relu(S)  # Only positive selection

        # Use torch.zeros_like to safely modify without inplace ops
        S_masked = torch.zeros_like(S)  # Create a mask to avoid in-place ops
        S_masked[..., 1:] = S[..., 1:]  # Do not mask <BOS> token, leave it unchanged

        eye_mask = 1 - torch.eye(T, device=S.device)  # Do not mask self
        S = S_masked * eye_mask  # Apply the masking to avoid self-attention

        S_shifted = torch.roll(S, 1, -2)  # Shift to mask strictly in the future
        S_shifted[..., 0, :] = 0  # Ensure future masking without inplace

        F_accum = torch.cumsum(S_shifted, dim=-2)  # Accumulate along sequence length
        att -= F_accum[:, None]  # Subtract accumulated attention from original logits

        att = F.softmax(att, dim=-1)

        y = att @ v
        y = y.transpose(1, 2).contiguous().view(B, T, C)
        y = self.c_proj(y)
        return y
```

**Comment by lucidrains** [comment]
you can have the previous layer control whether the next layer should have attention masked out, by doing a cumulative sum across depth dimension

**Comment by lucidrains** [comment]
I guess this all goes back to an old residual attention paper. perhaps the attention matrix should be kept around in a separate stream

**Comment by kazunator** [comment]
Oh that's super cool. Would love it if you could share the paper.
Also update: I was able to notice the perplexity gains shown by the paper with my implementation, which is pretty cool. The only weird thing is that selective attention uses almost double the memory of normal one because my implementation requires creating another matrix. I'll have to look through your logic and see if you're doing it differently

**Comment by lucidrains** [comment]
@kazunator yes that's right, they must have reused the first head so they could fine tune existing models, but i was thinking if this technique has wings, you'd just give an extra head to the queries and keys for this

**Comment by lucidrains** [comment]
> Oh that's super cool. Would love it if you could share the paper. Also update: I was able to notice the perplexity gains shown by the paper with my implementation, which is pretty cool. The only weird thing is that selective attention uses almost double the memory of normal one because my implementation requires creating another matrix. I'll have to look through your logic and see if you're doing it differently. Though I'll prolly have to follow your logic flow to understand what sim and sim_head actually are (I'm guessing sim head is the first head of the attention and you're just delegating that logic to an earlier part of the code)

residual attention is actually implemented in this repo, paper is [here](https://arxiv.org/abs/2012.11747)
